model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/InfinispanUserSessionProvider.java

```diff
@@ -270,8 +336,37 @@ private UserSessionEntity getUserSessionEntity(RealmModel realm, String id, bool
         return entity;
     }
 
+    private Stream<UserSessionModel> getUserSessionsFromPersistenceProviderStream(RealmModel realm, UserModel user, boolean offline) {
+        UserSessionPersisterProvider persister = session.getProvider(UserSessionPersisterProvider.class);
+        return persister.loadUserSessionsStream(realm, user, offline, 0, null)
+                .map(persistentUserSession -> getUserSessionEntityFromCacheOrImportIfNecessary(realm, offline, persistentUserSession))
+                .filter(Objects::nonNull)
+                .map(userSessionEntity -> wrap(realm, userSessionEntity, offline));
+    }
+
+
+    protected Stream<UserSessionModel> getUserSessionsStream(RealmModel realm, UserSessionPredicate predicate, boolean offline) {
+
+        if (offline && loadOfflineSessionsStatsFromDatabase) {
+
+            // fetch the offline user-sessions from the persistence provider
+            UserSessionPersisterProvider persister = session.getProvider(UserSessionPersisterProvider.class);
+
+            UserModel user = session.users().getUserById(realm, predicate.getUserId());
+            if (user != null) {
+                return persister.loadUserSessionsStream(realm, user, offline, 0, null);
+            }
+
+            // TODO add support for offline user-session lookup by brokerSessionId
```

Comment:
@hmlnarik Based on the [discussion here](https://github.com/keycloak/keycloak/pull/7722#issuecomment-770723908): 
which exception should I throw here?

Reply:
My vote is to throw ModelException as InfinispanUserSessionProvider is model implementation.

I am not 100% sure if we want to catch that exception in LogoutEndpoint and transform into some proper ErrorResponseException, or if we just don't handle exception? But IMO this is not user-related error (like for example updating password which doesn't match the password policy), but it is rather system error. Administrator should be aware that the use-case is not supported, so it will be ideal to make sure that it is shown in the server log and probably should be passed to the Event API as well.

Will be great if you have a chance to add automated test for this error scenario or at least check it manually. WDYT?

And I would throw the exception for both getOfflineUserSessionByBrokerSessionId and getOfflineUserSessionByBrokerUserId.

Reply:
Thanks for your input @mposolda !
I just adapted the appropriate method to throw the ModelException.